pennylane/beta/tapes/qubit_param_shift.py

```diff
@@ -0,0 +1,220 @@
+        self._evA = None
+        return super().jacobian(device, params, **options)
+
+    def parameter_shift(self, idx, device, params, **options):
+        r"""Partial derivative using the parameter-shift rule of a tape consisting of *only*
+        expectation values of observables.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Keyword Args:
+            shift (float): the parameter shift value
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        op = self._par_info[idx]["op"]
+        p_idx = self._par_info[idx]["p_idx"]
+
+        s = (
+            np.pi / 2
+            if op.grad_recipe is None or op.grad_recipe[p_idx] is None
+            else op.grad_recipe[p_idx]
+        )
+        s = options.get("shift", s)
+
+        shift = np.zeros_like(params)
+        shift[idx] = s
+
+        shift_forward = np.array(self.execute_device(params + shift, device))
+        shift_backward = np.array(self.execute_device(params - shift, device))
+
+        return (shift_forward - shift_backward) / (2 * np.sin(s))
+
+    def parameter_shift_var(self, idx, device, params, **options):
+        r"""Partial derivative using the parameter-shift rule of a tape consisting of a mixture
+        of expectation values and variances of observables.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (.Device, .QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): the quantum tape operation parameters
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        # Temporarily convert all variance measurements on the tape into expectation values
+        for i in self.var_idx:
+            self._measurements[i].return_type = qml.operation.Expectation
+
+        # Get <A>, the expectation value of the tape with unshifted parameters. This is only
+        # calculated once, if `self._evA` is not None.
+        if self._evA is None:
+            self._evA = np.asarray(self.execute_device(params, device))
+
+        # evaluate the analytic derivative of <A>
+        pdA = self.parameter_shift(idx, device, params, **options)
+
+        # For involutory observables (A^2 = I) we have d<A^2>/dp = 0.
+        # Currently, the only observable we have in PL that may be non-involutory is qml.Hermitian
+        involutory = [i for i in self.var_idx if self.observables[i].name != "Hermitian"]

```

Comment:
Will it always be the case that observables not called "Hermitian" are involutory?

Reply:
Probably not - if this ever changes, however, we will need to modify this method.

Currently the only observables we support are X, Y, Z, H, I, and qml.Hermitian. Of these, all are involutory (including tensor product combinations) except for qml.Hermitian.

Reply:
Perhaps, at some point, we should add a flag to the observables themselves, to mark them as involutory.

Reply:
yes